src/mlpack/tests/lsh_test.cpp

```diff
@@ -9,103 +9,211 @@
-
-  // After search
-  // COR.SOL.: 'neighbors' = [2 1 9; 3 8 2]
-  // COR.SOL.: 'distances' = [2 0 2; 4 2 16]
-
-  arma::Mat<size_t> true_neighbors(2, 3);
-  true_neighbors << 2 << 1 << 9 << arma::endr << 3 << 8 << 2 << arma::endr;
-  arma::mat true_distances(2, 3);
-  true_distances << 2 << 0 << 2 << arma::endr << 4 << 2 << 16 << arma::endr;
-
-  for (size_t i = 0; i < 3; i++)
-  {
-    for (size_t j = 0; j < 2; j++)
+  //read iris training and testing data as reference and query
+  const string data_train="iris_train.csv";
+  const string data_test="iris_test.csv";
+  arma::mat rdata;
+  arma::mat qdata;
+  data::Load(data_train, rdata, true);
+  data::Load(data_test, qdata, true);
+
+  //Run classic knn on reference data
+  AllkNN knn(rdata);
+  arma::Mat<size_t> groundTruth;
+  arma::mat groundDistances;
+  knn.Search(qdata, k, groundTruth, groundDistances);
+
+  //Test: Run LSH with varying number of tables, keeping all other parameters 
+  //constant. Compute the recall, i.e. the number of reported neighbors that
+  //are real neighbors of the query.
+  //LSH's property is that (with high probability), increasing the number of
+  //tables will increase recall. Epsilon ensures that if noise lightly affects
+  //the projections, the test will not fail.
+  //This produces false negatives, so we attempt the test numTries times and
+  //only declare failure if all of them fail.
+  
+  bool fail;
+
+  for (int t = 0; t < numTries; ++t){
+
+    fail = false;
+    const int Lsize = 6; //number of runs
+    const int L_value[] = {1, 8, 16, 32, 64, 128}; //number of tables
+    double L_value_recall[Lsize] = {0.0}; //recall of each LSH run
+
+    for (int l=0; l < Lsize; ++l)
     {
-//      BOOST_REQUIRE_EQUAL(neighbors(j, i), true_neighbors(j, i));
-//      BOOST_REQUIRE_CLOSE(distances(j, i), true_distances(j, i), 1e-5);
+      //run LSH with only numTables varying (other values default)
+      LSHSearch<> lsh_test1(rdata, numProj, L_value[l], 
+          hashWidth, secondHashSize, bucketSize);
+      arma::Mat<size_t> LSHneighbors;
+      arma::mat LSHdistances;
+      lsh_test1.Search(qdata, k, LSHneighbors, LSHdistances);
+
+      //compute recall for each query
+      L_value_recall[l] = compute_recall(LSHneighbors, groundTruth);
+
+      if (l > 0){
+        if(L_value_recall[l] < L_value_recall[l-1]-epsilon){
+          fail = true; //if test fails at one point, stop and retry
+          break;
+        }
+      }
     }
+    if ( !fail )
+    {
+      break; //if test passes one time, it is sufficient
+    }
+
+  }
+  BOOST_REQUIRE(fail == false);
+   
+  //Test: Run LSH with varying hash width, keeping all other parameters 
+  //constant. Compute the recall, i.e. the number of reported neighbors that
+  //are real neighbors of the query.
+  //LSH's property is that (with high probability), increasing the hash width
+  //will increase recall. Epsilon ensures that if noise lightly affects the 
+  //projections, the test will not fail.
```

Comment:
We should probably make this a separate `BOOST_AUTO_TEST_CASE()`, even though we'll have to load the data twice.